index.js

```diff
@@ -0,0 +1,116 @@
+const lazyloadOptions = {
+  rootElement: null,
+  rootMargin: `0px 0px ${window.innerHeight / 2}px 0px`,
+  treshold: 1,
+  lazyloadClass: 'loaded',
+  nativeLazyloadEnabled: 'loading' in HTMLImageElement.prototype,
+
+  getOptions() {
+    return {
+      root: this.rootElement,
+      rootMargin: this.rootMargin,
+      threshold: this.treshold
+    };
+  }
+};
+
+let lazyObserver = null;
+let resizeListener = null;
+let elements = [];
+let images = [];
+
+const setSource = element => {
+  const { src, srcset, wait } = element.dataset;
+
+  if (src) {
+    element.src = src;
+  }
+
+  if (srcset) {
+    element.srcset = srcset;
+  }
+
+  if (wait) {
+    element.onload = () => element.classList.add(lazyloadOptions.lazyloadClass);
+  }
+
+  element.removeAttribute('data-lazy');
+};
+
+const loadElements = () => {
+  elements = Array.from(document.querySelectorAll('audio[data-lazy], video[data-lazy], iframe[data-lazy], script[data-lazy]'));
+  images = Array.from(document.querySelectorAll('img[data-lazy]'));
+
+  if (lazyloadOptions.nativeLazyloadEnabled) {
+    images.forEach(image => {
+      image.loading = 'lazy';
+      setSource(image);
+    });
+  } else {
+    elements = elements.concat(images);
+  }
+};
+
+const onIntersect = (entries, observer) => {
+  entries.forEach(entry => {
+    if (entry.isIntersecting) {
+      const { target } = entry;
+
+      setSource(target);
+
+      observer.unobserve(target);
+    }
+  });
+};
+
+const destroyObserver = () => {
```

Comment:
This is a great function. I would love to go one step ahead and get rid of the observer if it's no longer needed. If we finish loading all, or all were native images, we don't need an observer anymore.

Reply:
Feel free to use promises or something like that. I feel like `Promise.all` would do wonders here without needing to keep count of things loading and things missing.